### DISCUSSION TITLE
feature proposal : alias for a group of classes

### DISCUSSION THREAD

**Discussion by salembaira** [tailwindlabs/tailwindcss#3344] [Ideas]
Hello guys, fantastic work!

Sometimes you end up reusing the same sequence of classes across your UI for divs or buttons for example. Reusable components can help but you may end up reusing the same set of classes across the reusable components too.

My proposition is to define your `aliases` inside `theme` like this :
```javascript
aliases: {
  "custom-div": "space-y-20 sm:space-y-32 md:space-y-40 2xl:space-y-44 overflow-hidden pt-6 px-4 sm:px-6 lg:px-8 mx-auto max-w-screen-2xl"
}
````

and inside your code do use it like this :
```html
<div className="aliases:custom-div">
  ...
</div>
````

What do you think?

**Comment by vovayatsyuk** [comment]
I think you can do the very same with `@apply` inside your css:

```css
.custom-div {
    @apply tailwind-classes
}
```

```html
<div className="custom-div">
  ...
</div>
```

**Comment by salembaira** [comment]
My bad, I didn't know @apply existed. It's exactly what I needed and more! Thanks ;)

**Comment by lianghx-319** [comment]
There are some different of my 「alias」.

## @apply case

### Input

```css
.custom-div {
    @apply w-full flex ...
}
```

```html
<div class="custom-div"><div>
```

### Output

```css
.custom-div {
    width: 100%;
    display: flex;
    ...
}
```

## My alias case

### Input

```js
// tailwind config
{
  alias: {"custom-div": "w-full flex ..."}
}
```

```html
<div class="custom-div"><div>
```

### Output

```css
.w-full {
    width: 100%;
}

.flex {
    display: flex;
}

...
```

```html
<div class="w-full flex ..."><div>
```

> This is a real reuse atomic css class, not just copy tailwind classes value into a custom class name

**Reply by vovayatsyuk** [reply]
Tailwind is generating CSS based on HTML.

The way you want it to work, you can write a `replace` tool that will take source HTML with "custom-class", create dist HTML with replaced "custom-clases" according to your rules.

**Comment by Detzler** [comment]
Hi @Baira @vovayatsyuk 
I stumbled across your discussion here and actually I currently have the same thought which you had two years ago. Instead of using apply which create duplicate css, it should be possible to use aliases that get replaced by some build tool or something.
How did you solve this issue at that time?

@vovayatsyuk @adamwathan Since everybody curse "apply" nowadays, do you know about any tools/solutions which has been created in the tailwind community to solve this issue? The alias feature by @Baira makes a lot of sence.
